File: src/main/java/org/oewntk/model/InverseRelationFactory.java
/*
 * Copyright (c) 2021-2021. Bernard Bou.
 */

package org.oewntk.model;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

/**
 * Generator of inverse synset relations
 */
public class InverseRelationFactory
{
	private static final boolean LOG_ALREADY_PRESENT = false;

	private static final Map<String, String> INVERSE_SYNSET_RELATIONS = new HashMap<>();

	static
	{
		INVERSE_SYNSET_RELATIONS.put("hypernym", "hyponym");
		INVERSE_SYNSET_RELATIONS.put("instance_hypernym", "instance_hyponym");
		INVERSE_SYNSET_RELATIONS.put("mero_part", "holo_part");
		INVERSE_SYNSET_RELATIONS.put("mero_member", "holo_member");
		INVERSE_SYNSET_RELATIONS.put("mero_substance", "holo_substance");
		INVERSE_SYNSET_RELATIONS.put("causes", "is_caused_by");
		INVERSE_SYNSET_RELATIONS.put("entails", "is_entailed_by");
		INVERSE_SYNSET_RELATIONS.put("exemplifies", "is_exemplified_by");
		INVERSE_SYNSET_RELATIONS.put("domain_topic", "has_domain_topic");
		INVERSE_SYNSET_RELATIONS.put("domain_region", "has_domain_region");
	}

	private static final Set<String> INVERSE_SYNSET_RELATIONS_KEYS = INVERSE_SYNSET_RELATIONS.keySet();

	private static final Map<String, String> INVERSE_SENSE_RELATIONS = new HashMap<>();

	static
	{
		INVERSE_SENSE_RELATIONS.put("exemplifies", "is_exemplified_by");
		INVERSE_SENSE_RELATIONS.put("domain_topic", "has_domain_topic");
		INVERSE_SENSE_RELATIONS.put("domain_region", "has_domain_region");
	}

	private static final Set<String> INVERSE_SENSE_RELATIONS_KEYS = INVERSE_SENSE_RELATIONS.keySet();

	private InverseRelationFactory()
	{
	}

	/**
	 * Generate inverse synset relations
	 *
	 * @param synsetsById synsets mapped by id
	 */
	public static int makeSynsetRelations(Map<String, Synset> synsetsById)
	{
		int count = 0;
		for (Map.Entry<String, Synset> entry : synsetsById.entrySet())
		{
			String sourceSynsetId = entry.getKey();
			Synset sourceSynset = entry.getValue();
			Map<String, Set<String>> relations = sourceSynset.getRelations();
			if (relations != null && relations.size() > 0)
			{
				for (String type : INVERSE_SYNSET_RELATIONS_KEYS)
				{
					Collection<String> targetSynsetIds = relations.get(type);
					if (targetSynsetIds != null && targetSynsetIds.size() > 0)
					{
						String inverseType = INVERSE_SYNSET_RELATIONS.get(type);
						for (String targetSynsetId : targetSynsetIds)
						{
							Synset targetSynset = synsetsById.get(targetSynsetId);
							assert targetSynset != null;
							try
							{
								targetSynset.addInverseRelation(inverseType, sourceSynsetId);
								count++;
							}
							catch (IllegalArgumentException e)
							{
								if (LOG_ALREADY_PRESENT)
								{
									Tracing.psErr.printf("[W] %s%n", e.getMessage());
								}
							}
						}
					}
				}
			}
		}
		return count;
	}

	/**
	 * Generate inverse sense relations
	 *
	 * @param sensesById senses mapped by id
	 */
	public static int makeSenseRelations(Map<String, Sense> sensesById)
	{
		int count = 0;
		for (Map.Entry<String, Sense> entry : sensesById.entrySet())
		{
			String sourceSenseId = entry.getKey();
			Sense sourceSense = entry.getValue();
			Map<String, Set<String>> relations = sourceSense.getRelations();
			if (relations != null && relations.size() > 0)
			{
				for (String type : INVERSE_SENSE_RELATIONS_KEYS)
				{
					Collection<String> targetSenseIds = relations.get(type);
					if (targetSenseIds != null && targetSenseIds.size() > 0)
					{
						String inverseType = INVERSE_SENSE_RELATIONS.get(type);
						for (String targetSenseId : targetSenseIds)
						{
							Sense targetSense = sensesById.get(targetSenseId);
							assert targetSense != null;
							try
							{
								targetSense.addInverseRelation(inverseType, sourceSenseId);
								count++;
							}
							catch (IllegalArgumentException e)
							{
								if (LOG_ALREADY_PRESENT)
								{
									Tracing.psErr.printf("[W] %s%n", e.getMessage());
								}
							}
						}
					}
				}
			}
		}
		return count;
	}
}

File: src/main/java/org/oewntk/model/InverseRelationFactory.kt
/*
 * Copyright (c) 2021-2024. Bernard Bou.
 */
package org.oewntk.model

import org.oewntk.model.Relation.Companion.INVERSE_SENSE_RELATIONS
import org.oewntk.model.Relation.Companion.INVERSE_SYNSET_RELATIONS

/**
 * Generator of inverse synset relations
 */
object InverseRelationFactory {

    private const val LOG_ALREADY_PRESENT = false

    private const val THROW_NON_EXISTING_TARGET = false

    val INVERSE_SYNSET_RELATIONS_MAP = INVERSE_SYNSET_RELATIONS.mapKeys { (k, _) -> Relation(k) }.mapValues { (_, v) -> Relation(v) }

    val INVERSE_SENSE_RELATIONS_MAP = INVERSE_SENSE_RELATIONS.mapKeys { (k, _) -> Relation(k) }.mapValues { (_, v) -> Relation(v) }

    val INVERSE_SYNSET_RELATIONS_SET = INVERSE_SYNSET_RELATIONS_MAP.values

    val INVERSE_SENSE_RELATIONS_SET = INVERSE_SENSE_RELATIONS_MAP.values

    /**
     * Generate inverse synset relations
     *
     * @param synsetsById synsets mapped by id
     * @param synsetsById synsets mapped by id
     * @return count
     */
    fun makeInverseSynsetRelations(synsetsById: Map<SynsetId, Synset>, sensesById: Map<SenseKey, Sense>): Int = makeInverseSynsetRelations(INVERSE_SYNSET_RELATIONS_MAP, synsetsById, sensesById)

    /**
     * Generate inverse synset relations as per map
     *
     * @param toInverse relation mapped to its inverse
     * @param synsetsById synsets mapped by id
     * @param sensesById senses mapped by id
     * @return count
     */
    fun makeInverseSynsetRelations(toInverse: Map<Relation, Relation>, synsetsById: Map<SynsetId, Synset>, sensesById: Map<SenseKey, Sense>): Int {
        var count = 0
        for ((sourceSynsetId, sourceSynset) in synsetsById) {
            if (!sourceSynset.relations.isNullOrEmpty()) {
                toInverse.keys.forEach {
                    val targetIds = sourceSynset.relations!![it]
                    if (!targetIds.isNullOrEmpty()) {
                        val inverseType = toInverse[it]
                        for (targetId in targetIds) {
                            if (targetId.targetsSynset) {
                                val targetSynset = synsetsById[targetId.synsetId]
                                if (targetSynset == null) {
                                    val message = "[E] non-existing target $targetId of synset relation $it($sourceSynsetId)"
                                    if (THROW_NON_EXISTING_TARGET) throw IllegalArgumentException(message) else {
                                        Tracing.psErr.println(message)
                                        continue
                                    }
                                }
                                try {
                                    targetSynset.addInverseRelation(inverseType!!, RelationTarget(sourceSynsetId.id))
                                    count++
                                } catch (e: IllegalArgumentException) {
                                    if (LOG_ALREADY_PRESENT) {
                                        Tracing.psErr.println("[W] ${e.message}\n")
                                    }
                                }
                            } else /* if (targetId.isSenseKey) */ {
                                val targetSense = sensesById[targetId.senseKey]
                                if (targetSense == null) {
                                    val message = "[E] non-existing target $targetId of synset relation $it($sourceSynsetId)"
                                    if (THROW_NON_EXISTING_TARGET) throw IllegalArgumentException(message) else {
                                        Tracing.psErr.println(message)
                                        continue
                                    }
                                }
                                try {
                                    targetSense.addInverseRelation(inverseType!!, RelationTarget(sourceSynsetId.id))
                                    count++
                                } catch (e: IllegalArgumentException) {
                                    if (LOG_ALREADY_PRESENT) {
                                        Tracing.psErr.println("[W] ${e.message}\n")
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
        return count
    }

    /**
     * Generate inverse sense relations
     *
     * @param sensesById senses mapped by id
     * @param synsetsById synsets mapped by id
     * @return count
     */
    fun makeInverseSenseRelations(sensesById: Map<SenseKey, Sense>, synsetsById: Map<SynsetId, Synset>): Int = makeInverseSenseRelations(INVERSE_SENSE_RELATIONS_MAP, sensesById, synsetsById)

    /**
     * Generate inverse sense relations
     *
     * @param toInverse relation mapped to its inverse
     * @param sensesById senses mapped by id
     * @param synsetsById synsets mapped by id
     * @return count
     */
    fun makeInverseSenseRelations(toInverse: Map<Relation, Relation>, sensesById: Map<SenseKey, Sense>, synsetsById: Map<SynsetId, Synset>): Int {
        var count = 0
        for ((sourceSenseId, sourceSense) in sensesById) {
            if (!sourceSense.relations.isNullOrEmpty()) {
                toInverse.keys.forEach {
                    val targetIds = sourceSense.relations!![it]
                    if (!targetIds.isNullOrEmpty()) {
                        val inverseType = toInverse[it]!!
                        for (targetId in targetIds) {
                            if (targetId.targetsSynset) {
                                val targetSynset = synsetsById[targetId.synsetId]
                                if (targetSynset == null) {
                                    val message = "[E] non-existing target $targetId of sense relation $it($sourceSenseId)"
                                    if (THROW_NON_EXISTING_TARGET) throw IllegalArgumentException(message) else {
                                        Tracing.psErr.println(message)
                                        continue
                                    }
                                }
                                try {
                                    targetSynset.addInverseRelation(inverseType, RelationTarget(sourceSenseId.id))
                                    count++
                                } catch (e: IllegalArgumentException) {
                                    if (LOG_ALREADY_PRESENT) {
                                        Tracing.psErr.println("[W] ${e.message}\n")
                                    }
                                }
                            } else /* if (targetId.isSenseKey) */ {
                                val targetSense = sensesById[targetId.senseKey]
                                if (targetSense == null) {
                                    val message = "[E] non-existing target $targetId of sense relation $it($sourceSenseId)"
                                    if (THROW_NON_EXISTING_TARGET) throw IllegalArgumentException(message) else {
                                        Tracing.psErr.println(message)
                                        continue
                                    }
                                }
                                try {
                                    targetSense.addInverseRelation(inverseType, RelationTarget(sourceSenseId.id))
                                    count++
                                } catch (e: IllegalArgumentException) {
                                    if (LOG_ALREADY_PRESENT) {
                                        Tracing.psErr.println("[W] ${e.message}\n")
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
        return count
    }
}
